### PR TITLE
fix: enable torch.autocast for TP parallelism without FSDP

### PR DESCRIPTION
**Pull request by eous**
Remove overly conservative restriction that disabled mixed precision for TP-only configurations. torch.autocast operates at the operator level and is orthogonal to tensor parallelism.

Before: TP-only training would show warning and disable mixed precision
After: TP-only training uses torch.autocast for mixed precision

Note: PP-only training uses schedule-based execution and doesn't use maybe_enable_amp (unchanged by this PR).

Affected configurations:
- TP-only (now enabled)
- DDP-only (was already enabled)
- Single-device (was already enabled)
- FSDP/HSDP (unchanged - handled internally by fully_shard)

### CODE REVIEW

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Have you verified it works properly? Could you show evidence, in terms of param / activation / grad dtype, and throughput comparison with mixed precision off?

I vaguely remember that I've tried it before and it didn't work as expected.

**Comment by eous** [comment]
> Have you verified it works properly? Could you show evidence, in terms of param / activation / grad dtype, and throughput comparison with mixed precision off?
> 
> I vaguely remember that I've tried it before and it didn't work as expected.
weight_diff_analysis_s3

https://huggingface.co/eousphoros/persona_eta_20b_131k This model was trained with TP=4 no fdsp. The output with autocast was inline with what I expected though I lack the depth of knowledge to formerly confirm this.

**Comment by tianyu-l** [comment]
Thanks. It's hard to tell from the plots that it is working properly.
My impression is that `autocast` is not actively maintained. Is it true? @fegin 

Also curious why would you use amp with TP but without FSDP?

**Comment by fegin** [comment]
> My impression is that autocast is not actively maintained. Is it true? @fegin

autocast is the core feature. But autocast with parallelisms is not actively maintained. We have seen performance gap between autocast + DDP and FSDP implementation with world_size being the same.  IMO, if we can use FSDP we should use FSDP.

**Comment by slabhs-aws** [comment]
@fegin @tianyu-l imo we should still support autocast for TP-only (no FSDP) workloads. During model bringup, we often start small with single core workloads, scale to TP, and then scale to FSDP. Accuracy validation at each point requires FP32, AMP, MXFP8 etc precisions supported. Ideally all precisions should support all parallelisms..

**Comment by tianyu-l** [comment]
@slabhs-aws I don't think people have plans working on autocast + TP.

We are baking some new idea, which I think would tackle the problem along the way. See https://github.com/pytorch/pytorch/compare/main...ailing/mst#diff-e326c855a215daccfd35d8953f8ce5f7c75cf2c446eb9880b22b512a8b1d7158

cc @ailzhang @weifengpy

**Comment by weifengpy** [comment]
> @slabhs-aws I don't think people have plans working on autocast + TP.
> 
> We are baking some new idea, which I think would tackle the problem along the way. See [pytorch/pytorch@main...ailing/mst#diff-e326c855a215daccfd35d8953f8ce5f7c75cf2c446eb9880b22b512a8b1d7158](https://github.com/pytorch/pytorch/compare/main...ailing/mst#diff-e326c855a215daccfd35d8953f8ce5f7c75cf2c446eb9880b22b512a8b1d7158)
> 
> cc @ailzhang @weifengpy

@tianyu-l This autocast happens after "mst.to_compute_state()" so I didn't understand your comment yet

**Comment by tianyu-l** [comment]
@weifengpy mst should handle mixed precision as FSDP2 does, so autocast is not needed.

I didn't know much on who's maintaining autocast. If you know, could you tag them here?

**Comment by weifengpy** [comment]
> @weifengpy mst should handle mixed precision as FSDP2 does, so autocast is not needed.

mst takes care of replicated parameters. for DDP or TP-only, autocast is still relevant?

> I didn't know much on who's maintaining autocast. If you know, could you tag them here?

@janeyx99 on autocast

**Comment by tianyu-l** [comment]
> mst takes care of replicated parameters. for DDP or TP-only, autocast is still relevant?

what is the difference between autocast and FSDP2 mixed precision? Since mst is not tied to DP and is about unifying storage, why mst wouldn't handle all params' dtype conversion? E.g. I imagine in TP, `mst.to_compute_state` would take fp32 shard param (e.g. colwise paralleled tensor), and returns a bf16 shard param for computation.

**Comment by weifengpy** [comment]
> what is the difference between autocast and FSDP2 mixed precision?

autocast does dtype cast op-by-op. fsdp mixed precision does dtype cast layer-by-layer (including input/output tensors). there is perf advantage when fsdp takes care of mixed precision

when MST is ready, it does dtype cast for its managed parameters. But input/output (activations) tensors are not managed by MST. We need pre/post forward/backward hooks to take care of dtype casting for input/output

**Comment by tianyu-l** [comment]
@weifengpy
fwiw:
- in LLM training, the input is integer. With FSDP we don't have to do this casting because the embedding layer is run in bf16 thanks to fsdp mixed precision handling.
- autocast doesn't support embedding, per https://github.com/pytorch/torchtitan/issues/1525

**Comment by weifengpy** [comment]
> * in LLM training, the input is integer.

good point! I missed this
